subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/DependencyInsightReportTask.java

```diff
@@ -206,37 +240,144 @@ public void renderNode(StyledTextOutput target, RenderableDependency node, boole
+                if (requested.equals(actualAttribute)) {
+                    if (actualValue.equals(requestedValue)) {
+                        return new AttributeMatchDetails(MatchType.requested, requested, requestedValue);
+                    }
+                    return new AttributeMatchDetails(MatchType.different_value, requested, requestedValue);
+                } else {
+                    // maybe it matched through coercion
+                    if (actualValue.toString().equals(requestedValue.toString())) {
+                        return new AttributeMatchDetails(MatchType.requested, requested, requestedValue);
+                    }
+                    return new AttributeMatchDetails(MatchType.different_value, requested, requestedValue);
                 }
+            }
+        }
+        return new AttributeMatchDetails(MatchType.not_requested, null, null);
+    }
 
-            }, true);
-            dependencyGraphRenderer.render(dependency);
-            boolean last = i++ == sortedDeps.size();
-            if (!last) {
-                output.println();
+    private static class AttributeMatchDetails {
+        private final MatchType matchType;
+        private final Attribute<?> requested;
+        private final Object requestedValue;
+
+        private AttributeMatchDetails(MatchType matchType, Attribute<?> requested, Object requestedValue) {
+            this.matchType = matchType;
+            this.requested = requested;
+            this.requestedValue = requestedValue;
+        }
+    }
+
+    private enum MatchType {
+        requested,
+        different_value,
+        not_requested
+    }
+
+    private static class RenderDependencyAction implements Action<StyledTextOutput> {
+        private final RenderableDependency dependency;
+        private final Configuration configuration;
+        private final boolean showVariantDetails;
+
+        public RenderDependencyAction(RenderableDependency dependency, Configuration configuration, boolean showVariantDetails) {
+            this.dependency = dependency;
+            this.configuration = configuration;
+            this.showVariantDetails = showVariantDetails;
+        }
+
+        public void execute(StyledTextOutput out) {
+            out.withStyle(Identifier).text(dependency.getName());
+            if (StringUtils.isNotEmpty(dependency.getDescription())) {
+                out.withStyle(Description).text(" (" + dependency.getDescription() + ")");
+            }
+            printVariantDetails(out);
+            switch (dependency.getResolutionState()) {
+                case FAILED:
+                    out.withStyle(Failure).text(" FAILED");
+                    break;
+                case RESOLVED:
+                    break;
+                case UNRESOLVED:
+                    out.withStyle(Failure).text(" (n)");
+                    break;
             }
+        }
 
+        private void printVariantDetails(StyledTextOutput out) {
```

Comment:
I think this would be more readable if you extract some more methods, like `printVariantAttribute()` or `printMissingAttributes()`. I find the method a bit too long and appears like a wall of text.